modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -97,19 +97,47 @@ object Sink {
                 }
     } yield exists
 
-  private def getRetryPolicy[F[_]: Applicative](config: BackoffPolicy): RetryPolicy[F] =
+  private def getRetryPolicyForErrors[F[_]: Applicative](config: BackoffPolicy): RetryPolicy[F] =
     capDelay[F](config.maxBackoff, fullJitter[F](config.minBackoff))
       .join(limitRetries(config.maxRetries))
 
+  private def getRetryPolicyForThrottling[F[_]: Applicative](config: BackoffPolicy): RetryPolicy[F] =
+    capDelay[F](config.maxBackoff, fullJitter[F](config.minBackoff))
```

Comment:
I wonder if we need a retry policy for throttling and if we shouldn't retry aggressively. The write quota that we reach is per second, and if we wait 2 seconds before retrying, we lost 2 seconds where we could have inserted records.

Reply:
Good discussion point!

It's a balance.... sending a batch to kinesis consumes some cpu and some network, both of which we should treat as scarce resources.  If we use cpu and network unnecessarily then we are stealing it away from the enrichment process.  By delaying a little bit we allow the enrichment process to have full access to cpu/network for a few more milliseconds.

Our current retry policy is like this:

- The 1st retry is delayed by (100ms * random factor for jitter) -- so 50 ms on average
- The 2nd retry is delayed by (2 * 100 ms * random factor) -- 100 ms on average
- The 5th retry is delayed by (16 * 100 ms * random factor) -- 800 ms on average
- The 10th retry is delayed by (512 * 100 ms * random factor) -- 25.6 seconds on average

(hope I've understood the policy correctly).

In reality I don't expect us to need too many retries; in production we scale up the number of shards to accommodate the event volume.

Now I look at those numbers... how about we keep the exponential backoff and jitter, but cap the max delay to ~1 second?

Reply:
> how about we keep the exponential backoff and jitter, but cap the max delay to ~1 second?

That sounds perfect !